PocketKit/Sources/PocketKit/Refresh/AbstractRefreshCoordinatorProtocol.swift

```diff
@@ -0,0 +1,166 @@
+    func handleSession(session: SharedPocketKit.Session?) {
+        guard let session = session else {
+            // If the session is nil, ensure the user's view is logged out
+            self.tearDownSession()
+            return
+        }
+
+        // We have a session! Ensure the user is logged in.
+        self.setUpSession(session)
+    }
+
+    /// Sets up the class with the logged in session
+    /// - Parameter session: The session we are setting up
+    private func setUpSession(_ session: SharedPocketKit.Session) {
+        notificationCenter.publisher(for: UIScene.didEnterBackgroundNotification, object: nil).sink { [weak self] _ in
+            guard let self else {
+                Log.captureNilWeakSelf()
+                return
+            }
+            self.submitRequest()
+        }.store(in: &subscriptions)
+
+        notificationCenter.publisher(for: UIScene.willEnterForegroundNotification, object: nil).sink { [weak self] _ in
+            guard let self else {
+                Log.captureNilWeakSelf()
+                return
+            }
+            self.refreshData {}
+        }.store(in: &subscriptions)
+
+        // The user just logged in, we need to refresh the latest data
+        self.refreshData {}
+
+        // The user just logged in, lets setup background refreshing
+        self.submitRequest()
+    }
+
+    /// Unsubscribes all listeners and cancels any pending background tasks
+    private func tearDownSession() {
+        subscriptions = []
+        taskScheduler.cancel(taskID)
+    }
+
+    /// Submit the request to be scheduled anytime after the given interval
+    private func submitRequest() {
+        guard let refreshInterval else {
+            Log.info("No refresh interval set by developer, not scheduling a refresh for \(type(of: self))")
+            return
+        }
+
+        guard appSession.currentSession != nil else {
+            Log.warning("No user session, so not scheduling a refresh \(type(of: self))")
+            return
+        }
+
+        do {
+            let request: BGTaskRequest
+            switch backgroundRequestType {
+            case .processing:
+                request = BGProcessingTaskRequest(identifier: taskID)
+                (request as? BGProcessingTaskRequest)?.requiresNetworkConnectivity = true
+            case .refresh:
+                request = BGAppRefreshTaskRequest(identifier: taskID)
+            }
+            request.earliestBeginDate = Date().addingTimeInterval(refreshInterval)
+            try taskScheduler.submit(request)
+        } catch {
+            Log.warning("Could not submit background task request for \(type(of: self))")
+            Log.capture(error: error)
+        }
+    }
+
+    /// Private function that calls the underlying refresh function
+    /// - Parameter completion: Completion to call when done refreshing data
+    private func refreshData(_ completion: @escaping () -> Void) {
```

Comment:
nit: we could make the closure optional if there are cases (like some above) where we don't want to pass one?

Reply:
i think we should always pass one but thats just me.